README
Name
    ganglia - distributed monitoring system

Version
    ganglia 3.3.5

    The latest version of this software and document will always be found at
    http://ganglia.sourceforge.net/.

Synopsis
         ______                  ___
        / ____/___ _____  ____ _/ (_)___ _
       / / __/ __ `/ __ \/ __ `/ / / __ `/
      / /_/ / /_/ / / / / /_/ / / / /_/ /
      \____/\__,_/_/ /_/\__, /_/_/\__,_/
                       /____/ Distributed Monitoring System

    Ganglia is a scalable distributed monitoring system for high-performance
    computing systems such as clusters and Grids. It is based on a
    hierarchical design targeted at federations of clusters. It relies on a
    multicast-based listen/announce protocol to monitor state within
    clusters and uses a tree of point-to-point connections amongst
    representative cluster nodes to federate clusters and aggregate their
    state. It leverages widely used technologies such as XML for data
    representation, XDR for compact, portable data transport, and RRDtool
    for data storage and visualization. It uses carefully engineered data
    structures and algorithms to achieve very low per-node overheads and
    high concurrency. The implementation is robust, has been ported to an
    extensive set of operating systems and processor architectures, and is
    currently in use on over 500 clusters around the world. It has been used
    to link clusters across university campuses and around the world and can
    scale to handle clusters with 2000 nodes.

    The ganglia system is comprised of two unique daemons, a PHP-based web
    frontend and a few other small utility programs.

    Ganglia Monitoring Daemon (gmond)
        Gmond is a multi-threaded daemon which runs on each cluster node you
        want to monitor. Installation is easy. You don't have to have a
        common NFS filesystem or a database backend, install special
        accounts, maintain configuration files or other annoying hassles.

        Gmond has four main responsibilities: monitor changes in host state,
        announce relevant changes, listen to the state of all other ganglia
        nodes via a unicast or multicast channel and answer requests for an
        XML description of the cluster state.

        Each gmond transmits in information in two different ways:
        unicasting/multicasting host state in external data representation
        (XDR) format using UDP messages or sending XML over a TCP
        connection.

    Ganglia Meta Daemon (gmetad)
        Federation in Ganglia is achieved using a tree of point-to-point
        connections amongst representative cluster nodes to aggregate the
        state of multiple clusters. At each node in the tree, a Ganglia Meta
        Daemon ("gmetad") periodically polls a collection of child data
        sources, parses the collected XML, saves all numeric, volatile
        metrics to round-robin databases and exports the aggregated XML over
        a TCP sockets to clients. Data sources may be either "gmond"
        daemons, representing specific clusters, or other "gmetad" daemons,
        representing sets of clusters. Data sources use source IP addresses
        for access control and can be specified using multiple IP addresses
        for failover. The latter capability is natural for aggregating data
        from clusters since each "gmond" daemon contains the entire state of
        its cluster.

    Ganglia PHP Web Frontend
        The Ganglia web frontend provides a view of the gathered information
        via real-time dynamic web pages. Most importantly, it displays
        Ganglia data in a meaningful way for system administrators and
        computer users. Although the web frontend to ganglia started as a
        simple HTML view of the XML tree, it has evolved into a system that
        keeps a colorful history of all collected data.

        The Ganglia web frontend caters to system administrators and users.
        For example, one can view the CPU utilization over the past hour,
        day, week, month, or year. The web frontend shows similar graphs for
        Memory usage, disk usage, network statistics, number of running
        processes, and all other Ganglia metrics.

        The web frontend depends on the existence of the "gmetad" which
        provides it with data from several Ganglia sources. Specifically,
        the web frontend will open the local port 8651 (by default) and
        expects to receive a Ganglia XML tree. The web pages themselves are
        highly dynamic; any change to the Ganglia data appears immediately
        on the site. This behavior leads to a very responsive site, but
        requires that the full XML tree be parsed on every page access.
        Therefore, the Ganglia web frontend should run on a fairly powerful,
        dedicated machine if it presents a large amount of data.

        The Ganglia web frontend is written in the PHP scripting language,
        and uses graphs generated by "gmetad" to display history
        information. It has been tested on many flavours of Unix (primarily
        Linux) with the Apache webserver and the PHP module (5.0.0 or
        later). The GD graphics library for PHP is used to generate pie
        charts in the frontend and needs to be installed separately. On
        RPM-based system, it is usually provided by the php-gd package.

Installation
    The latest version of all ganglia software can always be downloaded from
    http://ganglia.info/

    Ganglia runs on Linux (i386, x86_64, ia64, sparc, alpha, powerpc, m68k,
    mips, arm, hppa, s390), FreeBSD, NetBSD, OpenBSD, DragonflyBSD, MacOS X,
    Solaris, AIX, IRIX, Tru64, HPUX and Windows NT/XP/2000/2003/2008 making
    it as portable as it is scalable.

  Monitoring Core Installation
    If you use the Linux RPMs provided on the ganglia web site, you can skip
    to the end of this section.

    Ganglia uses the GNU autoconf so compilation and installation of the
    monitoring core is basically

      % ./configure
      % make
      % make install

    but there are some issues that you need to take a look at first.

    Kernel multicast support
        If you use the ganglia multicast support, you must have a kernel
        that supports multicast. The vast majority of machines have
        multicast support by default. If you have problems with ganglia this
        is a core issue.

    Gmetad is not installed by default
        Since "gmetad" relies on the Round-Robin Database Tool ( see
        http://www.rrdtool.org/ ) it will not be compiled unless you
        explicit request it by using a --with-gmetad flag.

          % ./configure --with-gmetad

        The configure script will fail if it cannot find the rrdtool library
        and header files. By default, it expects to find them at
        /usr/include/rrd.h and /usr/lib/librrd.so. If you installed them in
        different locations then you need to instruct configure where to
        find them using:

          % ./configure --with-librrd=/rrd/path --with-gmetad

        Of course, you need to substitute "/rrd/path" with the real location
        of the rrd tool directory where the header file can be located
        inside an include subdirectory and the library can be located inside
        a lib subdirectory. As an alternative you could set "-L" in LDFLAGS,
        and "-I" in CFLAGS and CPPFLAGS for the library path and the header
        path respectively.

    AIX should not be compiled with shared libraries
        You must add the "--disable-shared" configure flags if you are
        running on AIX. For more details refer to the README.AIX file

          % ./configure --disable-shared

    Solaris dependencies could be problematic
        Not really a Solaris specific problem, but since Solaris has several
        different package repositories, all of them unofficial, it is
        difficult to be sure that all possible permutations have been
        confirmed to work reliably.

        Be sure to have all dependencies covered, as explained in the
        INSTALL file and to use GNU make and a gcc compiler that builds
        32bit binaries with all other libraries matching that ISA.

        When in doubt, build the problematic dependency from source and
        remember to distribute it together with your ganglia build as
        everything is dynamically linked by default.

        Be particularly careful with libConfuse, especially if using the old
        2.5 version. LibConfuse 2.5 is known to be incorrectly packaged and
        to compile by default as a static library which will fail to link
        with ganglia.

    Propietary *NIX systems might not work at all
        The good news is that the libmetrics code that used to work before
        3.1 is still most likely working fine and so there is nothing
        fundamentally broken about it.

        But the bad news is that in order to add the dynamic metric
        functionality, the build system and the way gmond used to locate its
        metrics had to be changed significantly. Therefore getting gmond to
        build and work again required fixes to be implemented for all
        platforms.

        Since none of the developers had access to HPUX, IRIX, Tru64
        (OSF/1), or Darwin (MacOS X) those platforms might not be able to
        build or run a 3.1 gmond yet. If you have access to any of these
        platforms and want to run ganglia 3.1, feel free to drop by the
        ganglia-developers list with suggestions, or even better patches.

    GEXEC confusion
        GEXEC is a scalable cluster remote execution system which provides
        fast, RSA authenticated remote execution of parallel and distributed
        jobs. It provides transparent forwarding of stdin, stdout, stderr,
        and signals to and from remote processes, provides local environment
        propagation, and is designed to be robust and to scale to systems
        over 1000 nodes. Internally, GEXEC operates by building an n-ary
        tree of TCP sockets and threads between gexec daemons and
        propagating control information up and down the tree. By using
        hierarchical control, GEXEC distributes both the work and resource
        usage associated with massive amounts of parallelism across multiple
        nodes, thereby eliminating problems associated with single node
        resource limits (e.g., limits on the number of file descriptors on
        front-end nodes). (from http://www.theether.org/gexec )

        "gexec" is a great cluster execution tool but integrating it with
        ganglia is a bit clumsy. GEXEC can run standalone without access to
        a ganglia "gmond". In standalone mode gexec will use the hosts
        listed in your GEXEC_SVRS variable to run on. For example, say I
        want to run "hostname" on three machines in my cluster: "host1",
        "host2" and "host3". I use the following command line.

          % GEXEC_SVRS="host1 host2 host3" gexec -n 3 hostname

        and gexec would build an n-ary tree (binary tree by default) of TCP
        sockets to those machines and run the command "hostname"

        As an added feature, you can have "gexec" pull a host list from a
        locally running gmond and use that as the host list instead of
        GEXEC_SVRS. The list is load balanced and "gexec" will start the job
        on the *n* least-loaded machines.

        For example..

          % gexec -n 5 hostname

        will run the command "hostname" on the five least-loaded machines in
        a cluster.

        To turn on the "gexec" feature in ganglia you must configure ganglia
        with the "--enable-gexec" flag

          % ./configure --enable-gexec

        Enabling "gexec" means that by default any host running gmond will
        send a special message announcing that gexec is installed on it and
        open for requests.

        Now the question is, what if I don't want gexec to run on every host
        in my cluster? For example, you may not want to have "gexec" run
        jobs on your cluster frontend nodes.

        You simply add the following line to your "gmond" configuration file
        ("/etc/ganglia/gmond.conf" by default)

          no_gexec on

        Simple huh? I know the configuration file option, "no_gexec", seems
        crazy (and it is). Why have an option that says "yes to no gexec"?
        The early versions of gmond didn't use a configuration file but
        instead commandline options. One of the commandline options was
        simply "--no-gexec" and the default was to announce gexec as on.

    Once you have successfully run

      % ./configure <options>
      % make
      % make install

    you should find the following files installed in "/usr" (by default).

      /usr/bin/gstat
      /usr/bin/gmetric
      /usr/sbin/gmond
      /usr/sbin/gmetad

    If you installed ganglia using RPMs then these files will be installed
    when you install the RPM. The RPM is installed simply by running

      % rpm -Uvh ganglia-gmond-3.3.5.i386.rpm
      % rpm -Uvh ganglia-gmetad-3.3.5.i386.rpm

    Once you have the necessary binaries installed, you can test your
    installation by running

       % ./gmond

    This will start the ganglia monitoring daemon. You should then be able
    to run

       % telnet localhost 8649

    And get an XML description of the state of your machine (and any other
    hosts running gmond at the time).

    If you are installing by source on Linux, scripts are provided to start
    "gmetad" and "gmond" at system startup. They are easy to install from
    the source root.

       % cp ./gmond/gmond.init /etc/rc.d/init.d/gmond
       % chkconfig --add gmond
       % chkconfig --list gmond
         gmond              0:off   1:off   2:on    3:on    4:on    5:on    6:off
       % /etc/rc.d/init.d/gmond start
         Starting GANGLIA gmond:                                    [  OK  ]

    Repeat this step with gmetad.

  PHP Web Frontend Installation
    1.  The ./web directory of the ganglia distribution contains all the
        necessary PHP files for running your web frontend. Copy those files
        to "/var/www/html", however look for the variable "DocumentRoot" in
        your Apache configuration files to be sure. All the PHP script files
        use relative URLs in their links, so you may place the "ganglia/"
        directory anywhere convenient.

    2.  Ensure your webserver understands how to process PHP script files.
        Currently, the web frontend contains certain php language that
        requires PHP version 5 or greater. Processing PHP script files
        usually requires a webserver module, such as the "mod_php" for the
        popular Apache webserver. In RedHat Linux, the RPM package that
        provides this module is called simply "php".

        For Apache, "mod_php" module must be enabled. The following lines
        should appear somewhere in Apache's *conf files. This example
        applies to Red Hat Linux (and clones). The actual filenames may vary
        on your system. If you installed the php module using an RPM
        package, this work will have been done automatically.

          LoadModule php5_module modules/libphp5.so

          AddHandler php5-script .php
          AddType text/html .php

    3.  The webfrontend requires the existance of the gmetad package on the
        webserver. Follow the installation instructions on the gmetad page.
        Specifically, the webfrontend requires the rrdtool and the "rrds/"
        directory from gmetad. If you are a power user, you may use NFS to
        simulate the local existance of the rrds.

    4.  Test your installation. Visit the URL:

          http://localhost/ganglia/

        With a web-browser, where localhost is the address of your
        webserver.

    Installation of the web frontend is simplified on Linux by using rpm.

      % rpm -Uvh ganglia-web-3.3.5-1.noarch.rpm
      Preparing...                ########################################### [100%]
         1:ganglia-web            ########################################### [100%]

Configuration
  Gmond Configuration
    The configuration file format has changed between gmond version 2.5.x
    and version 3.x. The change was necessary in order to allow more complex
    configuration options.

    Gmond has a default configuration it will use if it does not find the
    default configuration file /etc/ganglia/gmond.conf. To see the default
    configuration simply run the command:

      % gmond --default_config

    and gmond will output its default configuration to stdout. This default
    configuration can serve as a good starting place for building a more
    custom configuration.

      % gmond --default_config > gmond.conf

    would create a file gmond.conf which you can then edit to taste and copy
    to /etc/ganglia/gmond.conf or elsewhere.

    To start gmond with a configuration file other then
    /etc/ganglia/gmond.conf, simply specify the configuration file location
    by running

      % gmond --config /my/ganglia/configs/custom.conf

    If you want to convert a 2.5.x configuration file to 3.x file format,
    run the following command

      % gmond --convert ./old_25_config.conf

    and gmond with output the equivalent 3.x configuration file to stdout.
    You can then redirect that output to a new configuration file which can
    serve as a starting point for your configuration.

      % gmond --convert ./old_25_config.conf > ./new_26_config.conf

    For details about gmond configuration options, simply run

      % man gmond.conf

    for a complete listing of options with detailed explanations.

  Gmetad Configuration
    The behavior of the Ganglia Meta Daemon is completely controlled by a
    single configuration file which is by default
    "/etc/ganglia/gmetad.conf". For gmetad to do anything useful you much
    specify at least one "data_source" in the configuration. The format of
    the data_source line is as follows

      data_source "Cluster A" 127.0.0.1  1.2.3.4:8655  1.2.3.5:8625
      data_source "Cluster B" 1.2.4.4:8655

    In this example, there are two unique data sources: "Cluster A" and
    "Cluster B". The Cluster A data source has three redundant sources. If
    gmetad cannot pull the data from the first source, it will continue
    trying the other sources in order.

    If you do not specify a port number, gmetad will assume the default
    ganglia port which is 8649 (U*N*I*X on a phone key pad)

    For a sample gmetad configuration file with comments, look at the
    gmetad.conf file provided as part of the distribution package in the
    gmetad directory

    "gmetad" has a "--conf" option to allow you to specify alternate
    configuration files

      % ./gmetad -conf=/tmp/my_custom_config.conf

  PHP Web Frontend Configuration
    Most configuration parameters reside in the "ganglia/conf.php" file.
    Here you may alter the template, gmetad location, RRDtool location, and
    set the default time range and metrics for graphs.

    The static portions of the Ganglia website are themable. This means you
    can alter elements such as section lables, some links, and images to
    suit your individual tastes and environment. The "template_name"
    variable names a directory containing the current theme. Ganglia uses
    TemplatePower to implement themes. A user-defined skin must conform to
    the template interface as defined by the default theme. Essentially, the
    variable names and START/END blocks in a custom theme must remain the
    same as the default, but all other HTML elements may be changed.

    Other configuration variables in "conf.php" specify the location of
    gmetad's files, and where to find the rrdtool program. These locations
    need only be changed if you do not run gmetad on the webserver.
    Otherwise the default locations should work fine. The "default_range"
    variable specifies what range of time to show on the graphs by default,
    with possible values of hour, day, week, month, year. The
    "default_metric" parameter specifies which metric to show on the cluster
    view page by default.

Commandline Tools
    There are two commandline tools that work with "gmond" to add custom
    metrics and query the current state of a cluster: "gmetric" and "gstat"
    respectively.

  Gmetric
    The Ganglia Metric Tool (gmetric) allows you to easily monitor any
    arbitrary host metrics that you like expanding on the core metrics that
    gmond measures by default.

    If you want help with the gmetric sytax, simply use the "help"
    commandline option

      % gmetric --help
      gmetric 3.3.5

      Purpose:
        The Ganglia Metric Client (gmetric) announces a metric
        on the list of defined send channels defined in a configuration file

      Usage: gmetric [OPTIONS]...

        -h, --help          Print help and exit
        -V, --version       Print version and exit
        -c, --conf=STRING   The configuration file to use for finding send channels
                            (default=`/etc/ganglia/gmond.conf')
        -n, --name=STRING   Name of the metric
        -v, --value=STRING  Value of the metric
        -t, --type=STRING   Either
                            string|int8|uint8|int16|uint16|int32|uint32|float|double
        -u, --units=STRING  Unit of measure for the value e.g. Kilobytes, Celcius
                            (default=`')
        -s, --slope=STRING  Either zero|positive|negative|both  (default=`both')
        -x, --tmax=INT      The maximum time in seconds between gmetric calls
                            (default=`60')
        -d, --dmax=INT      The lifetime in seconds of this metric  (default=`0')
        -S, --spoof=STRING  IP address and name of host/device (colon separated) we
                              are spoofing  (default='')
        -H, --heartbeat     spoof a heartbeat message (use with spoof option)

    Gmetric sends the metric specified on the commandline to all
    udp_send_channels specified in the configuration file
    /etc/ganglia/gmond.conf by default. If you want to send metric to
    alternate udp_send_channels, you can specify a different configuration
    file as such:

      % gmetric --conf=./custom.conf -n "wow" -v "it works" -t "string"

    All metrics in ganglia have a name, value, type and optionally units.
    For example, say I wanted to measure the temperature of my CPU
    (something gmond doesn't do by default) then I could send this metric
    with name="temperature", value="63", type="int16" and units="Celcius".

    Assume I have a program called "cputemp" which outputs in text the
    temperature of the CPU

      % cputemp
      63

    I could easily send this data to all listening gmonds by running

      % gmetric --name temperature --value `cputemp` --type int16 --units Celcius

    Check the exit value of gmetric to see if it successfully sent the data:
    0 on success and -1 on failure.

    To constantly sample this temperature metric, you just need too add this
    command to your cron table.

  Gstat

debian/changelog
ganglia (3.5.0-wm1) precise; urgency=low

  * Imported Upstream version 3.5.0
  * Applied diff by Nicholas Satterly

 -- Mark Bergsma <mark@wikimedia.org>  Tue, 12 Mar 2013 14:44:37 +0000

ganglia (3.3.8-2) UNRELEASED; urgency=low

  * Add a debian/watch file.
  * Stop removing debian/patches directory in clean target so that we
    can include patches to upstream.
  * Fix truncation of large block sizes in disk space reporting
    (Closes: Bug#694623).

 -- Mario Lang <mlang@debian.org>  Thu, 29 Nov 2012 09:32:58 +0100

ganglia (3.3.8-1) unstable; urgency=low

  * Check URL arguments thoroughly (Closes: #683584)
  * Fix un-initialized return code variable
  * Fix memory leak
  * Fix issue where Ganglia fails to start with NetworkManager
    (Closes: #638628)
  * Extra logging of buffer sizes for troubleshooting a common problem
  * Add Daniel Pocock as uploader (Debian Maintainer)
  * Add VCS URLs to control file.
  * PO translation didn't get into the package properly.

  [ Stuart Teasdale ]
  * Stop deleting web/version.php during clean as it comes from upstream

 -- Daniel Pocock <daniel@pocock.com.au>  Mon, 13 Aug 2012 15:17:28 +0200

ganglia (3.3.5-2) unstable; urgency=low

  * Add dutch translation. Closes: #658494.
  * Activated hardening flags from dpkg-buildflags. Closes: #655126.
  * Change pid file generation so the daemons do it. 
    Closes: #651293 #526146 #65124
  * Add httpd to the depends: for the web front end. Closes: #578620.
  * New translation for Polish. Closes: #667085.
  * New translation for Brazilian Portuguese. Closes: #667563.

 -- Stuart Teasdale <sdt@debian.org>  Thu, 05 Apr 2012 18:27:56 +0100

ganglia (3.3.5-1) experimental; urgency=low

  * New upstream version

 -- Stuart Teasdale <sdt@debian.org>  Tue, 27 Mar 2012 22:22:03 +0100

ganglia (3.3.1-1) unstable; urgency=low

  * New upstream release

 -- Stuart Teasdale <sdt@debian.org>  Thu, 08 Mar 2012 20:11:24 +0000


ganglia (3.2.0-1) unstable; urgency=low

  * New upstream release

 -- Stuart Teasdale <sdt@debian.org>  Sat, 21 Jan 2012 10:35:23 +0000


ganglia (3.1.7-2.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix "FTBFS: aborting due to unexpected upstream changes": use
    dh_autotools-dev_* helper tools for handling config.{guess,sub} in
    debian/rules, make build dependency on autotools-dev versioned, remove
    respective hunks from debian/patches/debian-changes-3.1.7-2.
    (Closes: #651320)
  * Now the package builds once but not twice in a row due to auto-generated
    files; clean them in debian/rules, and remove them from
    debian/patches/debian-changes-3.1.7-2.
  * Fix "post-installation script returns error exit status 1" and "post
    install script fails to add the ganglia user":
    patch from Ubuntu / Mark Mims:
    - Fix options on useradd in postinst. (LP: #854866)
    (Closes: #626407, #561199)

 -- gregor herrmann <gregoa@debian.org>  Tue, 10 Jan 2012 16:55:13 +0100
ganglia (3.1.7-2) unstable; urgency=low
  
  * Stop ganglia-monitor from changing permissions on /var/lib/ganglia
    in postinst. Closes: #586909
  * Switch from adduser to useradd in postinst and prerm. Closes: #620505 
  * Include Danish debconf template. Closes: #605328
  * Remove la files from dependency_libs. Closes: #621232 #620721
  * Switch to dpkg-source 3.0 (quilt) format

 -- Stuart Teasdale <sdt@debian.org>  Thu, 07 Apr 2011 21:26:32 +0100

ganglia (3.1.7-1) unstable; urgency=low

  * New upstream version. Closes: #584276. 
  * Bump of standards-version to 3.8.4 
  * Add build-dep on libpcre3-dev

 -- Stuart Teasdale <sdt@debian.org>  Tue, 15 Jun 2010 22:33:13 +0100

ganglia (3.1.2-3.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix pending l10n issues. Debconf translations:
    - Czech (Michal Simunek).  Closes: #538741
    - Italian (Vincenzo Campanella).  Closes: #556105
    - Czech (Michal Simunek).  Closes: #574010

 -- Christian Perrier <bubulle@debian.org>  Mon, 22 Mar 2010 00:45:10 +0100

ganglia (3.1.2-3) unstable; urgency=high

  * Adjust the default umask of the daemons. Closes: #567175
  * Fix in postinst to change the umask of existing rrds
  * Add build dependency for libdbi0-dev. Closes: #569395
  * Horrid directory mangling in postinst hacked around for permissions 

 -- Stuart Teasdale <sdt@debian.org>  Sat, 13 Feb 2010 23:15:02 +0000

ganglia (3.1.2-2.1) unstable; urgency=low

  * Non-maintainer upload by arrangement with maintainer.

  [ Jonathan Wiltshire ]
  * Debconf templates and debian/control reviewed by the debian-l10n-
    english team as part of the Smith review project. Closes: #539204
  * Debconf translation updates:
     - Swedish (closes: #539217)
     - Finnish (closes: #539325)
     - Czech (closes: #539614)
     - Basque (closes: #539872)
     - French (closes: #540038)
     - Italian (closes: #540200)
     - German (closes: #540220)
     - Spanish (closes: #540708)
     - Portugese (closes: #540769)
     - Russian (closes: #540849)
     - Vietnamese (closes: #547916)
     - Japanese (closes: #551022)
  * Refactor ganglia-webfrontend.{post,pre}inst for changed templates, and
    simultaneously purge old code now that apache2 is the only package in
    Squeeze. Use the a2enmod and a2dismod helpers to enable and disable the
    Apache module instead of directly creating symbolic links.

  [ Christian Perrier ]
  * Debconf translation updates:
    - Galician (Marce Villarino).  Closes: #555451

 -- Christian Perrier <bubulle@debian.org>  Tue, 10 Nov 2009 22:02:43 +0100

ganglia (3.1.2-2) unstable; urgency=low

  * Tweaks to ganglia-webfrontend debconf templates
  * If we have an old gmond.conf convert it. Closes: #512154 
  * Changes to the example apache aliases. Closes: #518255

 -- Stuart Teasdale <sdt@debian.org>  Thu, 02 Jul 2009 19:59:35 +0100

ganglia (3.1.2-1) experimental; urgency=low

  * New upstream version with security fixes. (Closes: #512365.)
  * Include the gmond.conf manpage in the ganglia-monitor package.
    Closes: #512695.

gmetad/xml_hash.c
/* C code produced by gperf version 3.0.3 */
/* Command-line: gperf -l -H xml_hash -t -F ', 0' -N in_xml_list -k '1,$' -W xml_tags xml_hash.gperf  */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gnu-gperf@gnu.org>."
#endif

#line 1 "xml_hash.gperf"

/* $Id: xml_hash.c 2619 2011-06-27 19:47:46Z vvuksan $ */
#include <gmetad.h>
#line 5 "xml_hash.gperf"
struct xml_tag;

#define TOTAL_KEYWORDS 33
#define MIN_WORD_LENGTH 2
#define MAX_WORD_LENGTH 13
#define MIN_HASH_VALUE 4
#define MAX_HASH_VALUE 49
/* maximum key range = 46, duplicates = 0 */

#ifdef __GNUC__
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
xml_hash (str, len)
     register const char *str;
     register unsigned int len;
{
  static unsigned char asso_values[] =
    {
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50,  0, 50,  5, 25,  5,
      50,  0, 35, 25, 50, 50, 10,  5,  5,  5,
      10, 50, 10,  0,  0, 20, 15, 50, 15, 40,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50, 50, 50, 50, 50,
      50, 50, 50, 50, 50, 50
    };
  return len + asso_values[(unsigned char)str[len - 1]] + asso_values[(unsigned char)str[0]];
}

#ifdef __GNUC__
__inline
#ifdef __GNUC_STDC_INLINE__
__attribute__ ((__gnu_inline__))
#endif
#endif
struct xml_tag *
in_xml_list (str, len)
     register const char *str;
     register unsigned int len;
{
  static unsigned char lengthtable[] =
    {
       0,  0,  0,  0,  4,  0,  0,  2,  3,  4,  5,  6,  7,  3,
       4, 10,  6,  7, 13,  4,  5, 11,  7,  8,  9,  5,  0,  7,
       3,  4,  0,  0,  2,  3,  4,  0,  0,  2, 13,  4,  5,  0,
       0,  8,  4,  0,  0,  0,  0,  9
    };
  static struct xml_tag xml_tags[] =
    {
      {"", 0}, {"", 0}, {"", 0}, {"", 0},
#line 11 "xml_hash.gperf"
      {"TAGS", TAGS_TAG},
      {"", 0}, {"", 0},
#line 31 "xml_hash.gperf"
      {"TN", TN_TAG},
#line 28 "xml_hash.gperf"
      {"SUM", SUM_TAG,},
#line 35 "xml_hash.gperf"
      {"TYPE", TYPE_TAG},
#line 36 "xml_hash.gperf"
      {"SLOPE", SLOPE_TAG},
#line 37 "xml_hash.gperf"
      {"SOURCE", SOURCE_TAG},
#line 27 "xml_hash.gperf"
      {"METRICS", METRICS_TAG},
#line 29 "xml_hash.gperf"
      {"NUM", NUM_TAG,},
#line 10 "xml_hash.gperf"
      {"NAME", NAME_TAG},
#line 38 "xml_hash.gperf"
      {"EXTRA_DATA", EXTRA_DATA_TAG},
#line 26 "xml_hash.gperf"
      {"METRIC", METRIC_TAG},
#line 16 "xml_hash.gperf"
      {"LATLONG", LATLONG_TAG,},
#line 39 "xml_hash.gperf"
      {"EXTRA_ELEMENT", EXTRA_ELEMENT_TAG},
#line 32 "xml_hash.gperf"
      {"TMAX", TMAX_TAG},
#line 15 "xml_hash.gperf"
      {"OWNER", OWNER_TAG,},
#line 7 "xml_hash.gperf"
      {"GANGLIA_XML", GANGLIA_XML_TAG},
#line 13 "xml_hash.gperf"
      {"CLUSTER", CLUSTER_TAG},
#line 23 "xml_hash.gperf"
      {"LOCATION", LOCATION_TAG},
#line 14 "xml_hash.gperf"
      {"LOCALTIME", LOCALTIME_TAG},
#line 30 "xml_hash.gperf"
      {"UNITS", UNITS_TAG},
      {"", 0},
#line 8 "xml_hash.gperf"
      {"VERSION", VERSION_TAG},
#line 34 "xml_hash.gperf"
      {"VAL", VAL_TAG},
#line 9 "xml_hash.gperf"
      {"GRID", GRID_TAG},
      {"", 0}, {"", 0},
#line 20 "xml_hash.gperf"
      {"UP", UP_TAG,},
#line 17 "xml_hash.gperf"
      {"URL", URL_TAG,},
#line 21 "xml_hash.gperf"
      {"DOWN", DOWN_TAG,},
      {"", 0}, {"", 0},
#line 22 "xml_hash.gperf"
      {"IP", IP_TAG},
#line 25 "xml_hash.gperf"
      {"GMOND_STARTED", STARTED_TAG},
#line 18 "xml_hash.gperf"
      {"HOST", HOST_TAG},
#line 19 "xml_hash.gperf"
      {"HOSTS", HOSTS_TAG,},
      {"", 0}, {"", 0},
#line 24 "xml_hash.gperf"
      {"REPORTED", REPORTED_TAG},
#line 33 "xml_hash.gperf"
      {"DMAX", DMAX_TAG},
      {"", 0}, {"", 0}, {"", 0}, {"", 0},
#line 12 "xml_hash.gperf"
      {"AUTHORITY", AUTHORITY_TAG}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register int key = xml_hash (str, len);

      if (key <= MAX_HASH_VALUE && key >= 0)
        if (len == lengthtable[key])
          {
            register const char *s = xml_tags[key].name;

            if (*str == *s && !memcmp (str + 1, s + 1, len - 1))
              return &xml_tags[key];
          }
    }
  return 0;
}
